Add selectConcat method to query builder

bennothommo opened the pull request:
Adds a new `selectConcat` method to the October Rain query builder, that takes an array of concatenation parts and makes them aliased to a specific field in the model. This will use the correct type of concatentation based on the database type, ie. CONCAT() for MySQL, || for other types.

Commit:
Add October DB Capsule Manager

Commit:
Add selectConcat method for query builder

Replaces Laravel's Grammar classes with October Rain versions, and sets the DB connections to use these new Grammars.

Commit:
Add unit test

Commit:
Code quality fixes

Commit:
Fix string handling

LukeTowers commented on src/Database/Query/Grammars/Grammar.php:
```diff
@@ -0,0 +1,94 @@
+        // If the query is actually performing an aggregating select, we will let that
+        // compiler handle the building of the select clauses, as it will need some
+        // more syntax that is best handled by that function to keep things neat.
+        if (! is_null($query->aggregate)) {
```
```suggestion
        if (!is_null($query->aggregate)) {
```

LukeTowers commented:
@bennothommo when this is polished up can you also submit a PR to Laravel? If they accept it into Laravel 6 then I'd like to just wait for us to move to 6 so that the maintenance burden is on them.

multiwebinc commented:
> Ready for review. The only thing I'm not sure of is this line:
> 
> https://github.com/octobercms/library/pull/434/files#diff-654762bea79e1fd19fbc793ef9990585R85
> 
> The idea is that column names should be wrapped in column quotes, and everything else is interpreted as a string, however, I couldn't find a method in the Builder scope that works out whether a given string is a column name. If anyone has any ideas on how that should be properly done, please feel free to point me in the correct direction.

This was what I was running into as well, which is why I was thinking quoted quotes. The problem with the current regex is that both strings and column names can contain practically any characters (including upper-case for column names) when wrapped correctly with quotes/backticks.

Here are 3 solutions I can think of:

- Match for `/^\'.*\'$/`. If the string starts and ends with a single quote, treat it as a string, or else treat it as a column.
- Use multi-dimensional arrays:

        $query = $model
            ->newQuery()
            ->select(['id'])
            ->selectConcat([
                [
                    'field',
                    'type' => 'identifier',
                ],
                [
                    ' ',
                    'type' => 'string',
                ],
                [
                    'cast'
                    'type' => 'identifier',
                ]
                'as' => 'full_cast'
            ]);

- Append each string with the type:

        $query = $model
            ->newQuery()
            ->select(['id'])
            ->selectConcat(['field|i', ' |s', 'cast|i'], 'full_cast');

Also, for both DB types, I would recommend quoting strings with single quotes and column names with backticks for MySQL and double quotes for SQLite. See [here](https://stackoverflow.com/questions/1992314/what-is-the-difference-between-single-and-double-quotes-in-sql).

multiwebinc commented on tests/Database/SelectConcatTest.php:
```diff
@@ -0,0 +1,53 @@
+            $query->toSql()
+        );
+    }
+}
```
Consider testing strings and identifiers in upper/lower case, as well as identifiers using reserved keywords (e.g. `select "order" from tablename`).

bennothommo commented:
@LukeTowers do you want me to refactor our implementation and proceed with including it in the Rain library? I have a feeling that Laravel are not interested in it.

LukeTowers commented:
@bennothommo sure, but hold off on merging it just yet

Commit:
Update src/Database/Query/Grammars/Grammar.php

Commit:
Merge branch 'develop' into wip/select-concat

Commit:
Simplify concat grammar overrides by using a trait

Instead of copying over the entire grammars for the different database types, we'll just replace and amend what is necessary, so we should be safe with updates.

Commit:
Store concat'd columns in query builder

Commit:
Update tests

Commit:
Fix code smells

Commit:
Wrap string literals with single quotes

Seems to be the most supported format across database servers.

Commit:
Add PostgreSQL and SQL Server tests

Commit:
Ensure multiple calls to selectConcat are included in query

LukeTowers commented:
@bennothommo could you add this to our DB docs?

bennothommo commented:
@LukeTowers done - https://github.com/octobercms/docs/commit/3d8ab788b355bd4a150d56113ac0ada6d7bcd863
